gcs_light_client.mediawiki

```diff
@@ -0,0 +1,1056 @@
+miners, we require an alternative (albeit less-binding) method to allow light
+clients to ''identify'' and ''reject'' invalid filters. The purely p2p solution
+to this problem is to obtain a deterministic hash-chain of ''each'' filter.
+This hash chain or "filter header chain" is similar to the regular Bitcoin
+headers in that it allows a light client to verify the ''authenticity'' of a
+received
+filter.
+
+The filter header chain for a particular filter type is described by the
+following recurrence:
+<pre>
+filter_header(n: uint) -> [32]byte = 
+   // The zero hash is 32 bytes of 0's.
+   let zero_hash [32]byte = {0*32}
+
+   if n == 0:
+       double-sha-256(genesis_block.prevblock || filter(0))
+
+   match filter(n): 
+      // If the filter isn't empty, then we hash the filter itself into the
+      // header chain.
+      case Some:
+          double-sha-256(filter_header(n-1) || double-sha-256(filter(n)))
+
+      // Otherwise, if the filter is empty (created from a block with a single
+      // coinbase transaction whose output script contains no push datas), then
+      // we'll hash the zero_hash.
+      case None:
+          double-sha-256(filter_header(n-1) || double-sha-256(zero_hash))
+
+   where filter(n) is the filter for block height n
+</pre>
+
+The filter header for the genesis block uses the hash stored in the prevblock
+field of the genesis block header itself, as there's no prior filter header
+(by definition).
+
+Due to the nature of filter construction, it's possible to construct a block
+such that an "empty" filter will be produced. This is the case of a coinbase
+transaction that has no data pushes in its output script. In this case, the
+"hash" of said filter is simply "32 zeroes". 
+
+
+We now introduce two new messages to support the fetching and verification of
+the filter header chain by light clients. 
+
+The <code>getcfheaders</code> message is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| Var-Int
+| NumBlockLocators
+| uint64
+| Number of block locators.
+|-
+| NumBlockLocators * 32
+| BlockLocatorHashes
+| [NumBlockLocators][32]byte
+| Block locator hashes, with the same semantics as in <code>getheaders</code>.
+|-
+| 32
+| HashStop
+| [32]byte
+| Hash to stop at.
+|-
+| 1
+| FilterType
+| byte
+| Type of filter header being requested.
+|}
+
+The <code>BlockLocators</code> within the message are to be interpreted
+identically to the <code>BlockLocators</code> within Bitcoin's
+<code>getheaders</code> and <code>getblocks</code> messages <ref>https://en.bitcoin.it/wiki/Protocol_documentation</ref>.
+
+The <code>cfheaders></code> message MUST be sent in response to a
+<code>getcfheaders</code> message for a particular block hash. The
+<code>cfheaders</code> message is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Descriptions
+! Data Type
+! Comments
+|-
+| 32
+| StopHash
```

Comment:
This is racy - lets say you request some filter headers (eg from a peer just finishing re-syncing after restart), so the request ends with a block hash that you dont have the header for, but the network converges on a different block...now you have a filter chain ending in a block that you dont have a header for and possibly cant get, but needlessly so. Switching things to backwards download (eg making getcfheaders take a block hash and then a number of blocks *before* that block to return filter headers for) should make this much less of an issue. Or, ideally, just make it a getcfheader request instead of getcfheaders, simplifes a lot of things and you can still do a backwards walk with a binary search to find the conflicting header, something you'll want to do either way.